### PR TITLE
test: re-enable hamt-test in interop

### PR DESCRIPTION
**Pull request by SgtPooki**
Based on fix from https://github.com/ipfs/helia/pull/448, only including e0f8d5e791082b15661e7276bf241d907bd5f59f & 3c400806f4dcaaf7cd6e4ed5ef7aa87673c1d57e, and then copying over some code from `ipfs-unixfs-exporter`

## Description

Re-enables the hamt-sharded test since it's reliably passing before 60 second timeout now (roughly 40seconds on my machine)

## Notes & open questions

Until the fix in https://github.com/ipfs/helia/pull/448 is sufficient, this should not be merged. We should decide on the proper fix, and utilize this test to ensure things are working as expected.

## Change checklist

- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation if necessary (this includes comments as well)
- [ ] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: packages/unixfs/src/commands/utils/find-shard-cid.ts [hunk 1]
```diff
@@ -0,0 +1,147 @@
+import { decode, type PBLink, type PBNode } from '@ipld/dag-pb'
```
Comment (SgtPooki) [inline packages/unixfs/src/commands/utils/find-shard-cid.ts:1]: all of this was copied over from `ipfs-unixfs-exporter`

File: packages/unixfs/src/commands/utils/resolve.ts [hunk 2]
```diff
@@ -4,6 +4,8 @@ import { DoesNotExistError, InvalidParametersError } from '../../errors.js'
 import { addLink } from './add-link.js'
```
Comment (SgtPooki) [inline packages/unixfs/src/commands/utils/resolve.ts:4]: All of the changes in here are originally from @aschmahmann in https://github.com/ipfs/helia/pull/448, except that I copied the contents of `./find-shard-cid.js` from `ipfs-unixfs-exporter`

File: packages/interop/src/verified-fetch-unixfs-dir.spec.ts [hunk 3]
```diff
@@ -76,16 +77,22 @@ describe('@helia/verified-fetch - unixfs directory', () => {
...
-      const resp = await verifiedFetch('ipfs://bafybeidbclfqleg2uojchspzd4bob56dqetqjsj27gy2cq3klkkgxtpn4i/685.txt')
+      const onProgress = sinon.stub()
+      const resp = await verifiedFetch('ipfs://bafybeidbclfqleg2uojchspzd4bob56dqetqjsj27gy2cq3klkkgxtpn4i/685.txt', { onProgress })
       expect(resp).to.be.ok()
       const text = await resp.text()
+      const onProgressEvents = onProgress.getCalls().map(call => call.args[0])
+      const walkEvents = onProgressEvents.filter((e) => e.type.includes('unixfs:exporter:walk'))
+      const blockGetEvents = onProgressEvents.filter((e) => e.type === 'blocks:get:providers:get')
+      expect(blockGetEvents).to.have.length(7)
+      expect(walkEvents).to.have.length.lessThanOrEqual(8)
```
Comment (SgtPooki) [inline packages/interop/src/verified-fetch-unixfs-dir.spec.ts:95]: note that I re-enabled this test on a branch https://github.com/ipfs/helia/tree/test/re-enable-hamt-on-main and it doesn't seem to fail there (it was only failing locally previously), but the events don't show the issues that are seen in https://github.com/ipfs-shipyard/helia-service-worker-gateway/issues/19 & https://github.com/ipfs-shipyard/helia-service-worker-gateway/issues/18.

They both have `onProgressEvents.length === 41` (which includes blockstore & verified fetch events), so some event is not being surfaced fully from the walk without the changes here.

**Review by SgtPooki — COMMENTED** [review]
self review

**Review by SgtPooki — COMMENTED** [review]
reviewing the latest changes (addition of sinon dep to interop)

File: packages/unixfs/package.json [hunk 4]
```diff
@@ -164,6 +164,7 @@
     "@libp2p/interface": "^1.1.2",
     "@libp2p/logger": "^4.0.5",
     "@multiformats/murmur3": "^2.1.8",
+    "err-code": "^3.0.1",
```
Comment (achingbrain) [inline packages/unixfs/package.json:167]: Please use `CodeError` from `@libp2p/interface`

**Comment by SgtPooki** [comment]
Closing in favor of #455

**Pull request closed without merging**